bolt-sidecar/src/state/pricing.rs

```diff
@@ -0,0 +1,271 @@
+        let inclusion_tip_ether = expected_val / (incoming_gas as f64);
+        let inclusion_tip_wei = (inclusion_tip_ether * 1e18) as u64;
+
+        Ok(inclusion_tip_wei)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_min_priority_fee_zero_preconfirmed() {
+        let pricing = PreconfPricing::default();
+
+        // Test minimum fee (21k gas ETH transfer, 0 preconfirmed)
+        let incoming_gas = 21_000;
+        let preconfirmed_gas = 0;
+        let min_fee_wei =
+            pricing.calculate_min_priority_fee(incoming_gas, preconfirmed_gas).unwrap();
+
+        // Pricing model article expects fee of 0.61 Gwei
+        assert!(
+            (min_fee_wei as f64 - 613_499_092.0).abs() < 1_000.0,
+            "Expected ~613,499,092 Wei, got {} Wei",
+            min_fee_wei
+        );
+    }
+
+    #[test]
+    fn test_min_priority_fee_zero_big_preconfirmed() {
+        let pricing = PreconfPricing::default();
+
+        // Test minimum fee (210k gas ETH transfer, 0 preconfirmed)
+        let incoming_gas = 210_000;
+        let preconfirmed_gas = 0;
+        let min_fee_wei =
+            pricing.calculate_min_priority_fee(incoming_gas, preconfirmed_gas).unwrap();
+
+        // This preconf uses 10x more gas than the 21k preconf,
+        // but the fee is only slightly higher.
+        assert!(
+            (min_fee_wei as f64 - 615_379_171.0).abs() < 1_000.0,
+            "Expected ~615,379,171 Wei, got {} Wei",
+            min_fee_wei
+        );
+    }
+
+    #[test]
+    fn test_min_priority_fee_medium_load() {
+        let pricing = PreconfPricing::default();
+
+        // Test medium load (21k gas, 15M preconfirmed)
+        let incoming_gas = 21_000;
+        let preconfirmed_gas = 15_000_000;
+        let min_fee_wei =
+            pricing.calculate_min_priority_fee(incoming_gas, preconfirmed_gas).unwrap();
+
+        // Pricing model article expects fee of ~1.17 Gwei
+        assert!(
+            (min_fee_wei as f64 - 1_189_738_950.0).abs() < 1_000.0,
+            "Expected ~1,189,738,950 Wei, got {} Wei",
+            min_fee_wei
+        );
+    }
+
+    #[test]
+    fn test_min_priority_fee_max_load() {
+        let pricing = PreconfPricing::default();
+
+        // Test last preconfirmed transaction (21k gas, almost 30M preconfirmed)
+        let incoming_gas = 21_000;
+        let preconfirmed_gas = 30_000_000 - 21_000;
+        let min_fee_wei =
+            pricing.calculate_min_priority_fee(incoming_gas, preconfirmed_gas).unwrap();
+
+        // Expected fee: ~19 Gwei
+        // This will likely never happen, since you want to reserve some gas
+        // on top of the block for MEV, but enforcing this is not the responsibility
+        // of the pricing model.
+        assert!(
+            (min_fee_wei as f64 - 19_175_357_339.0).abs() < 1_000.0,
+            "Expected ~19,175,357,339 Wei, got {} Wei",
+            min_fee_wei
+        );
+    }
```

Comment:
I'd add a disclaimer that these checks are mainly for tracking down whether some changes in the algo are highly different compared to previous result, but these target values are not to be considered "correct" since they have been produced with this function in the beginning.